zcash_client_backend/src/welding_rig.rs

```diff
@@ -81,30 +79,81 @@ fn scan_output<P: consensus::Parameters>(
             to,
             is_change,
             witness,
-            nf: Some(nf),
+            nf,
         });
     }
+
     None
 }
 
-/// Scans a [`CompactBlock`] with a set of [`ViewingKey`]s.
+pub trait ScanningKey {
+    type Nf;
+
+    fn try_decryption<P: consensus::Parameters>(
+        &self,
+        params: &P,
+        height: BlockHeight,
+        epk: &jubjub::ExtendedPoint,
+        cmu: &bls12_381::Scalar,
+        ct: &[u8],
+    ) -> Option<(Note, PaymentAddress)>;
+
+    fn nf(&self, note: &Note, witness: &IncrementalWitness<Node>) -> Self::Nf;
+}
+
+impl ScanningKey for ViewingKey {
+    type Nf = Nullifier;
+
+    fn try_decryption<P: consensus::Parameters>(
+        &self,
+        params: &P,
+        height: BlockHeight,
+        epk: &jubjub::ExtendedPoint,
+        cmu: &bls12_381::Scalar,
+        ct: &[u8],
+    ) -> Option<(Note, PaymentAddress)> {
+        try_sapling_compact_note_decryption(params, height, &self.ivk(), &epk, &cmu, &ct)
+    }
+
+    fn nf(&self, note: &Note, witness: &IncrementalWitness<Node>) -> Self::Nf {
+        note.nf(self, witness.position() as u64)
+    }
+}
+
+impl ScanningKey for SaplingIvk {
+    type Nf = ();
+
+    fn try_decryption<P: consensus::Parameters>(
+        &self,
+        params: &P,
+        height: BlockHeight,
+        epk: &jubjub::ExtendedPoint,
+        cmu: &bls12_381::Scalar,
+        ct: &[u8],
+    ) -> Option<(Note, PaymentAddress)> {
+        try_sapling_compact_note_decryption(params, height, self, &epk, &cmu, &ct)
+    }
+
+    fn nf(&self, _note: &Note, _witness: &IncrementalWitness<Node>) {}
+}
+
+/// Scans a [`CompactBlock`] with a set of [`ScanningKeys`]s.
 ///
 /// Returns a vector of [`WalletTx`]s belonging to any of the given
-/// [`ViewingKey`]s.
+/// [`ScanningKey`]s. If scanning with a full viewing key, the nullifiers
+/// of the resulting [`WalletShieldedOutput`]s will also be computed.
 ///
 /// The given [`CommitmentTree`] and existing [`IncrementalWitness`]es are
 /// incremented appropriately.
-///
-/// [`ExtendedFullViewingKey`]: zcash_primitives::zip32::ExtendedFullViewingKey
```

Comment:
Removing this doclink without replacing it causes CI errors. Either add a link to `ScanningKey`, or replace references to them with `ExtendedFullViewingKey or SaplingIvk`.